src/internal/grpc/middlewares-interceptor.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  InterceptingCall,
+  Interceptor,
+  InterceptorOptions,
+  Listener,
+  Metadata,
+  Requester,
+  StatusObject,
+} from '@grpc/grpc-js';
+import {NextCall} from '@grpc/grpc-js/build/src/client-interceptors';
+import {
+  Middleware,
+  MiddlewareRequestHandler,
+} from '../../config/middleware/middleware';
+import {MomentoLoggerFactory} from '../../config/logging/momento-logger';
+import {Message} from 'google-protobuf';
+
+export function middlewaresInterceptor(
+  loggerFactory: MomentoLoggerFactory,
+  middlewares: Middleware[]
+): Interceptor {
+  return (options: InterceptorOptions, nextCall: NextCall) => {
+    const middlewareRequestHandlers = middlewares.map(m => m.onNewRequest());
+    const requester: Requester = {
+      start: function (
+        metadata: Metadata,
+        listener: Listener,
+        next: (metadata: Metadata, listener: Listener) => void
+      ): void {
+        const newListener: Listener = {
+          onReceiveMetadata: function (
+            metadata: Metadata,
+            next: (metadata: Metadata) => void
+          ): void {
+            applyMiddlewareHandlers(
+              'onResponseMetadata',
+              middlewareRequestHandlers.reverse(),
```

Comment:
what is this reverse() for?

Reply:
when I'm composing the functions I need to do it in reverse order so that they will get applied in the order the user expects; i.e. the first middleware they pass needs to be the last one that I wrap with in order for it to get executed first.  I will double-check this but that's what I recall from my earlier testing.

Reply:
It seems unnatural to me to invert the order - middlewares conceptually "wrap" successively around one another. I'd expect the last thing I added to be the furthest from my application / closest to the wire. (and conversely, the first thing I added I'd expect to be right next to my application)
Is there some pre-existing pattern you're basing this on or some user feedback that they want reversed-wrap ordering instead of the normal grpc style layered ordering? Or maybe I don't understand and this reverse() is necessary to achieve the normal layered ordering?

Reply:
i'm saying that I believe this is how to get it to work the way you are describing, because of how I am composing the functions later.

if I add two middlewares, one that logs "FOO" before the request and one that logs "BAR" before the request, and I add them as `withMiddlewares[logFoo, logBar]`, then intuitively I expect to see the FOO log before the BAR log.  and i'm saying that this is how the code needs to be structured to make the function composition work that way.

as long as we are aligned that "FOO" should be logged before "BAR", i will doublecheck and make 100% sure that this code is yielding that behavior, but I'm almost certain that it is :)

Reply:
yes, "before the request" logging, I think it should be foobar. For _response_ event logging, I expect barfoo for that withMiddlewares you pasted. If it's just a symmetry issue somewhere downstream structurally then okay this is fine and I'm just confused by reading .reverse() here.

Reply:
okay, we are aligned on the expected behavior, and I believe that this PR does it exactly that way, but I will double check one more time before merging it and leave a comment on this line to clarify that point after i confirm.

Reply:
And works for me too! Thanks for confirming the expected vs. actual behavior!